### PR TITLE
Keep manual_kernel_registration only effective in aten codegen.

### PR DESCRIPTION
**Pull request by ailzhang**
This PR removes manual registration in aten/native codebase. 
And it separates manual device/catchall kernel registration from manual VariableType kernel registration. 
The first one remains as manual_kernel_registration in native_functions.yaml. 
The second one is moved to tools/ codegen.  

Difference in generated TypeDefault.cpp: https://gist.github.com/ailzhang/897ef9fdf0c834279cd358febba07734
No difference in generated VariableType_X.cpp

### CODE REVIEW

**Review by smessmer — COMMENTED** [review]
This seems to add more special cases to codegen where before the codegen was simple and the special cases were expressed by having C++ code handle it. I think it's preferable to have simple codegen and potentially more complex C++, but it's not a strong opinion. What do you think? cc @ezyang @bhosmer

**Comment by ailzhang** [comment]
@smessmer One downside of special casing in C++ is that it's hidden for downstream and we only know what to do by looking at the code. (Plus we won't be notified at API level if that code change) :P

**Comment by ezyang** [comment]
> This seems to add more special cases to codegen where before the codegen was simple and the special cases were expressed by having C++ code handle it. I think it's preferable to have simple codegen and potentially more complex C++, but it's not a strong opinion.

This seems like a non-sequitur. We're making a policy decision about whether or not autograd codegen is generating suitable VariableType kernels or not; previously, this policy decision was done compactly by specifying `manual_kernel_registration` on various kernels, but the compactness here was an illusion; as we established in discussion, the flag conflated two concepts: whether or not to generate registrations in ATen codegen, and whether or not to generate VariableType kernels.

Perhaps you are expressing a preference that we have another field in `native_functions.yaml`, named, say, `manual_variable_type_registration`. I will not die if this is the course of action we take, but I prefer the localized blacklist in `gen_variable_type.py` as it acknowledges that whether or not to generate variable type registration is *specifically* the business of `gen_variable_type.py`, and not some global, universally useful property that (1) ATen codegen could use and (2) that should be published in Declarations.yaml. Putting it in the codegen file is more encapsulated, and that's why I asked Ailing to do it this way.

**Review by smessmer — APPROVED** [review]

**Review by bhosmer — APPROVED** [review]
LGTM! thanks for the interactive run-through @ailzhang 😛

**Comment by ailzhang** [comment]
For reference:
Before this PR when we say `manual_kernel_registration` is true for an op, theoretically it has 4 possibilities:
1. op want to skip both registration to CPU/CUDA/catchall and Autograd/Trace. 
   `backward/set_data/data/is_leaf/output_nr/_version/requires_grad_/retain_grad`
2. op only want to skip registration to Autograd/Trace, but we still have to manually register it to catchAll. 
   `copy_/resize_as_/detach/detach_`
3. op only want to skip registration to Autograd/Trace, but we still have to manually register it to CPU/CUDA.
   `resize_`
4. op only want to skip registration to CPU/CUDA/catchall. 
   Currently there's no such use case in our codebase. 

After this PR, we do
1. set ops in (1) as `manual_kernel_registration=True` in aten codegen. 
2. put ops in (1) (2) (3) in blacklist in tools/ codegen. 
   - (1) => MANUAL_CATCHALL
   - (2) => MANUAL_AUTOGRAD_AND_TRACER

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -679,6 +705,9 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
                 wrapper_registrations.append(UNBOXEDONLY_WRAPPER_REGISTRATION.substitute(
                     declaration, class_type='VariableType'))
 
+        # See Note [Manual catchAll kernels]
+        assert (declaration['name'] in MANUAL_CATCHALL) == declaration['manual_kernel_registration']
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:709]: Why do you need the extra `MANUAL_CATCHALL` set if it has to exactly match the ones with `declaration['manual_kernel_registration'] == True` ?

Comment (ailzhang) [reply tools/autograd/gen_variable_type.py:709]: sorry for late reply! For the ops in MANUAL_CATCHALL we used to skip both aten & variable registration (so that catchall can be used). 
For more context see https://github.com/pytorch/pytorch/pull/42386#issuecomment-668274333. Note that (4) is possible but we currently don't have a use case for it yet so I added the assert so that we don't accidentally change the behavior. In the future if there's a use case we could remove the assert as needed.
